spark/src/main/java/com/netflix/iceberg/spark/source/Reader.java

```diff
@@ -265,21 +282,35 @@ private Schema lazyExpectedSchema() {
         .impl(UnsafeProjection.class, InternalRow.class)
         .build();
 
-    private final Iterator<FileScanTask> tasks;
     private final Schema tableSchema;
     private final Schema expectedSchema;
     private final FileIO fileIo;
 
+    private final Iterator<FileScanTask> tasks;
+    private final Map<String, InputFile> inputFiles;
+
     private Iterator<InternalRow> currentIterator = null;
     private Closeable currentCloseable = null;
     private InternalRow current = null;
 
-    public TaskDataReader(CombinedScanTask task, Schema tableSchema, Schema expectedSchema, FileIO fileIo) {
+    public TaskDataReader(CombinedScanTask task, Schema tableSchema, Schema expectedSchema, FileIO fileIo,
+                          EncryptionManager encryptionManager) {
       this.fileIo = fileIo;
       this.tasks = task.files().iterator();
       this.tableSchema = tableSchema;
       this.expectedSchema = expectedSchema;
       // open last because the schemas and fileIo must be set
+      Iterable<InputFile> inputFiles = encryptionManager.decrypt(() -> task.files().stream()
+          .map(fileScanTask ->
+              EncryptedFiles.encryptedInput(
+                  this.fileIo.newInputFile(fileScanTask.file().path().toString()),
+                  fileScanTask.file().keyMetadata()))
+          .iterator());
+      this.inputFiles = StreamSupport.stream(inputFiles.spliterator(), false)
```

Comment:
Minor: There isn't a problem with this, but I find it awkward compared to using `Iterables.transform` and a map builder:

```java
      Iterable<InputFile> decrypts = encryptionManager.decrypt(Iterables.transform(task.files(),
          fileTask -> EncryptedFiles.encryptedInput(
              fileIo.newInputFile(fileTask.file().path().toString()),
              fileTask.file().keyMetadata())));

      ImmutableMap.Builder<String, InputFile> builder = ImmutableMap.builder();
      decrypts.forEach(in -> builder.put(in.location(), in));
      this.inputFiles = builder.build();
```

In particular, using a lambda to create an `Iterable` out of an `Iterator` in order to convert back from `Stream`. And then using `StreamSupport` from a `spliterator` to munge the types enough to use `Collectors.toMap`. These stream APIs just don't seem to fit particularly well with each other or with the existing `Iterator` and `Iterable` interfaces.

This works well enough, but I would prefer it a little simpler, without streams.